[CSS] Matched element should be within the @scope

mdubet opened the pull request:
#### ee0320f9080b7d19f60aee912ac8acd10464f418
<pre>
[CSS] Matched element should be within the @scope
<a href="https://bugs.webkit.org/show_bug.cgi?id=265241">https://bugs.webkit.org/show_bug.cgi?id=265241</a>
<a href="https://rdar.apple.com/118713295">rdar://118713295</a>

Reviewed by Antti Koivisto.

This patch adds one part of the @scope matching algorithm: the matched element
has to be within the scope.

<a href="https://drafts.csswg.org/css-cascade-6/#scoped-styles">https://drafts.csswg.org/css-cascade-6/#scoped-styles</a>

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/scope-at-rule-expected.html:
* LayoutTests/fast/css/scope-at-rule.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-deep-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-evaluation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-focus-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-hover-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-implicit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-nesting-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-proximity-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-specificity-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-visited-cssom-expected.txt:
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::matchUARules):
(WebCore::Style::ElementRuleCollector::ruleMatches):
(WebCore::Style::ElementRuleCollector::scopeRulesMatch):
(WebCore::Style::ElementRuleCollector::matchAllRules):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::resolveSelectorListWithNesting):
(WebCore::Style::RuleSetBuilder::addStyleRule):
* Source/WebCore/style/RuleSetBuilder.h:

Canonical link: <a href="https://commits.webkit.org/271104@main">https://commits.webkit.org/271104@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef883a2572e3efd86bd060220edbe77d077977d3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27330 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5969 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28578 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29556 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25012 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27798 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7874 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3368 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24822 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27592 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4764 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23457 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4148 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4285 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24454 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30192 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24963 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24873 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30437 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4315 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2459 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28391 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5779 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/24189 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6585 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4772 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4696 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -584,14 +584,65 @@ bool ElementRuleCollector::containerQueriesMatch(const RuleData& ruleData, const
-    // https://bugs.webkit.org/show_bug.cgi?id=265241
-    return false;
+    SelectorChecker checker(element().rootElement()->document());
+    auto context = SelectorChecker::CheckingContext(SelectorChecker::Mode::ResolvingStyle);
```
why is checker initialized using one syntax and the context other?

mdubet replied:
No specific reason. It seems we don't have a coding style guideline for this.
Will use the non auto syntax for those.

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -584,14 +584,65 @@ bool ElementRuleCollector::containerQueriesMatch(const RuleData& ruleData, const
-    // https://bugs.webkit.org/show_bug.cgi?id=265241
-    return false;
+    SelectorChecker checker(element().rootElement()->document());
+    auto context = SelectorChecker::CheckingContext(SelectorChecker::Mode::ResolvingStyle);
```
why is the checker initialized using one syntax and the context other?

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -584,14 +584,65 @@ bool ElementRuleCollector::containerQueriesMatch(const RuleData& ruleData, const
+            selector.setSelectorList(makeUnique<CSSSelectorList>(WTFMove(scopeStart)));
+            if (!ancestorsMatch(selector))
+                return false;
+        }
```
Copying and constructing selectors in middle of style resolution is surely not the way to go? Why can't we match the scopeStart() selector list directly?

mdubet replied:
The behaviour described in spec is purposely similar to `:is()`  (ignore pseudo-element) and `matchSelectorList` is private in `SelectorChecker` and needs a `LocalContext` so it seems more direct to wrap it in `:is`. 

Should we try to adjust `SelectorChecker::matchSelectorList` a bit for this ?

anttijk replied:
You can just loop over the selectors and call match(). One of the existing SelectorChecker modes likely gives you the right behavior and if not you can add a new one.

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -584,14 +584,59 @@ bool ElementRuleCollector::containerQueriesMatch(const RuleData& ruleData, const
+            auto ancestor = element().parentElement();
+            while (ancestor) {
+                if (ancestor == olderAncestorToMatch) {
+                    // We limit the search to find an end of the donut to children of the start of the donut, not an ancestor
```
Can we not talk about donuts? No one outside CSSWG knows what that is about.

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -584,14 +584,59 @@ bool ElementRuleCollector::containerQueriesMatch(const RuleData& ruleData, const
+    SelectorChecker::CheckingContext context(SelectorChecker::Mode::CollectingRulesIgnoringVirtualPseudoElements);
+
+    auto evaluate = [&](auto& rule) {
+        Element* olderAncestorToMatch = nullptr;
```
"Older" is not the best way to describe this. This seems to be the scope start element, not sure if there is a spec term.
This can be const.

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -584,14 +584,59 @@ bool ElementRuleCollector::containerQueriesMatch(const RuleData& ruleData, const
+
+    // We need to respect each nested @scope to collect this rule
+    for (auto& rule : rules) {
+        if (!evaluate(rule))
```
Can this be called something more descriptive like `isWithinScope`?

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -584,14 +584,59 @@ bool ElementRuleCollector::containerQueriesMatch(const RuleData& ruleData, const
 bool ElementRuleCollector::scopeRulesMatch(const RuleData& ruleData, const MatchRequest& matchRequest)
 {
-    auto queries = matchRequest.ruleSet.scopeRulesFor(ruleData);
+    auto rules = matchRequest.ruleSet.scopeRulesFor(ruleData);
```
Lets call this `scopeRules`, as plain `rule` in ElementRuleCollector refers to style rules.

mdubet committed:
[CSS] Matched element should be within the @scope
https://bugs.webkit.org/show_bug.cgi?id=265241
rdar://118713295

Reviewed by Antti Koivisto.

This patch adds one part of the @scope matching algorithm: the matched element
has to be within the scope.

https://drafts.csswg.org/css-cascade-6/#scoped-styles

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/scope-at-rule-expected.html:
* LayoutTests/fast/css/scope-at-rule.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-deep-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-evaluation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-focus-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-hover-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-implicit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-nesting-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-proximity-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-specificity-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-visited-cssom-expected.txt:
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::matchUARules):
(WebCore::Style::ElementRuleCollector::ruleMatches):
(WebCore::Style::ElementRuleCollector::scopeRulesMatch):
(WebCore::Style::ElementRuleCollector::matchAllRules):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::resolveSelectorListWithNesting):
(WebCore::Style::RuleSetBuilder::addStyleRule):
* Source/WebCore/style/RuleSetBuilder.h:

Canonical link: https://commits.webkit.org/271104@main
